src/freenet/client/async/SplitFileInserterStorage.java

```diff
@@ -218,7 +223,16 @@
 
     private final int overallStatusLength;
     
+    /** Protects cooldown calculations and noBlocksToSend. You MUST NOT take this lock while 
+     * holding any other locks!!!
```

Comment:
Yes, the whole request queueing structure including the getters is strictly in-RAM, so we can change from per-splitfile getters to per-segment getters easily. It won't cost a lot of memory as the segment (and the bloom filter for the segment) is IIRC much bigger than the getter.

What I am referring to is all the complexity I tried to explain above. Selecting blocks at the level of a splitfile involves a lot of ugly unnecessary code (SplitFileFetcherStorage.chooseBlock()) that duplicates functionality elsewhere (the actual queue e.g. RandomGrabArray), is inefficient (read the above method!) and involves some very messy locking. Moving to per-segment getters would eliminate a lot of unnecessary code, reduce CPU usage, and avoid messy nested locking.